### PR TITLE
Staggered scheme for TM

### PR DESCRIPTION
**Pull request by wenqing**
As titled.

Two benchmarks are included, which including reference  results are from the existing two benchmarks for the monolithic scheme.

**Reference file**
- `tm_q_quad_pcs_1_ts_20_t_20000.000000.vtu` is a copy of `Tests/Data/ThermoMechanics/expected_tm_q_quad_pcs_0_ts_20_t_20000.000000.vtu`
- `CreepAfterExcavation_pcs_1_ts_61_t_4320000.000000.vtu` is a copy of ` Tests/Data/ThermoMechanics/CreepBGRa/CreepAfterExcavation/ExpectedCreepAfterExcavation_pcs_0_ts_61_t_4320000.000000.vtu `

### CODE REVIEW

File: ProcessLib/ThermoMechanics/ThermoMechanicsFEM.h [hunk 1]
```diff
@@ -659,6 +736,12 @@ class ThermoMechanicsLocalAssembler
     SecondaryData<typename ShapeMatrices::ShapeType> _secondary_data;
     bool const _is_axially_symmetric;
 
+    /// ID of the mechanical process.
+    int const _mechanics_process_id;
+
+    /// ID of heat conduction process.
+    int const _heat_conduction_process_id;
```
Comment (endJunction) [inline ProcessLib/ThermoMechanics/ThermoMechanicsFEM.h]: Better to store the process ids in process data, for example...

File: ProcessLib/ThermoMechanics/ThermoMechanicsProcess.cpp [hunk 2]
```diff
@@ -114,18 +118,35 @@ bool ThermoMechanicsProcess<DisplacementDim>::isLinear() const
...
+{
+    // Note: the heat conduction process and the mechnical process use the same
+    // order of shape functions.
+
+    // Create single component dof in every of the mesh's nodes.
+    _mesh_subset_all_nodes =
+        std::make_unique<MeshLib::MeshSubset>(_mesh, _mesh.getNodes());
 
-    // TODO move the two data members somewhere else.
-    // for extrapolation of secondary variables
```
Comment (endJunction) [inline ProcessLib/ThermoMechanics/ThermoMechanicsProcess.cpp]: Keep comment, please.

File: ProcessLib/ThermoMechanics/ThermoMechanicsProcess.cpp [hunk 3]
```diff
@@ -134,6 +155,65 @@ void ThermoMechanicsProcess<DisplacementDim>::initializeConcreteProcess(
...
+        // Collect the mesh subsets in a vector for each variable components for
+        // the mechanical process.
+        std::generate_n(std::back_inserter(all_mesh_subsets),
+                        _process_variables[_mechanics_process_id][0]
+                            .get()
+                            .getNumberOfComponents(),
+                        [&]() { return *_mesh_subset_all_nodes; });
+
+        // Create a vector of the number of variable components.
+        vec_var_n_components.push_back(
```
Comment (endJunction) [inline ProcessLib/ThermoMechanics/ThermoMechanicsProcess.cpp]: For discussion: Why only one push_back for mechanics and no for T?

File: ProcessLib/ThermoMechanics/ThermoMechanicsProcess.cpp [hunk 4]
```diff
@@ -201,6 +281,28 @@ void ThermoMechanicsProcess<DisplacementDim>::initializeConcreteProcess(
     }
 }
 
+template <int DisplacementDim>
+void ThermoMechanicsProcess<DisplacementDim>::initializeBoundaryConditions()
+{
+    if (_use_monolithic_scheme)
+    {
+        const int process_id_of_thermomechancs = 0;
```
Comment (endJunction) [inline ProcessLib/ThermoMechanics/ThermoMechanicsProcess.cpp]: This value was already stored; see the create process code, where the decision monolithic/staggered is made...

File: ProcessLib/ThermoMechanics/ThermoMechanicsFEM.h (hunk 1 above)
Comment (wenqing) [reply ProcessLib/ThermoMechanics/ThermoMechanicsFEM.h]: Moved.

File: ProcessLib/ThermoMechanics/ThermoMechanicsProcess.cpp (hunk 2 above)
Comment (wenqing) [reply ProcessLib/ThermoMechanics/ThermoMechanicsProcess.cpp]: Recovered.

File: ProcessLib/ThermoMechanics/ThermoMechanicsProcess.cpp (hunk 3 above)
Comment (wenqing) [reply ProcessLib/ThermoMechanics/ThermoMechanicsProcess.cpp]: Process::constructDofTable() handles the cases that all processes of a problem has the same number of primary variable (or the same DOF per node in FEM), which means all processes use a single  DOF table.

For TM, T and M have different DOF tables. Therefore  Process::constructDofTable() is overloaded in   ThermoMechanicsProcess.
DOF table:
T: _local_to_global_index_map_single_component, a member of ThermoMechanicsProcess.
M: _local_to_global_index_map, a member of Process.

File: ProcessLib/ThermoMechanics/ThermoMechanicsProcess.cpp (hunk 4 above)
Comment (wenqing) [reply ProcessLib/ThermoMechanics/ThermoMechanicsProcess.cpp]: Yes, you are right. For monolithic scheme, the process ID is 0, and the stored values are that  mechanics_process_id =0 and  heat_conduction_process_id =0. For this case, using either  mechanics_process_id =0 or  heat_conduction_process_id =0 would be misleading. Therefore this local constant is declared.

File: ProcessLib/ThermoMechanics/ThermoMechanicsProcess.cpp (hunk 3 above)
Comment (wenqing) [reply ProcessLib/ThermoMechanics/ThermoMechanicsProcess.cpp]: Process::constructDofTable() can be refactored later on to reduce code duplication in  the overloaded member of it in HM, PM. The refactored  Process::constructDofTable() can initialize Process::_local_to_global_index_map for all process, and the overloaded constructDofTable() can call Process::constructDofTable() and initialize the specific DOF table   of the derived classes.

File: ProcessLib/ThermoMechanics/ThermoMechanicsProcess.cpp (hunk 4 above)
Comment (endJunction) [reply ProcessLib/ThermoMechanics/ThermoMechanicsProcess.cpp]:
```suggestion
        const int process_id_of_thermomechanics = 0;
```

File: ProcessLib/ThermoMechanics/ThermoMechanicsFEM-impl.h [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+        // dV = 3 * alpha * dT * V_0
+        // rho_s_{n+1} = rho_s_n / (1 + 3 * alpha * dT )
+        // see reference solid density description for details.
+        auto& rho_s = _ip_data[ip].solid_density;
+        // calculate thermally induced strain
+        // assume isotropic thermal expansion
+        auto const alpha = _process_data.linear_thermal_expansion_coefficient(
+            t, x_position)[0];
+
+        double const dT = N.dot(local_dT);
```
Comment (endJunction) [inline ProcessLib/ThermoMechanics/ThermoMechanicsFEM-impl.h]: I'd think in other processes the standard notation was `dT_ip` for integration point value, and `dT` for part of `x_dot`. Same for the primary variables `T_ip` and `T`.

File: ProcessLib/ThermoMechanics/ThermoMechanicsFEM.h [hunk 6]
```diff
@@ -668,3 +741,5 @@ class ThermoMechanicsLocalAssembler
 
 }  // namespace ThermoMechanics
 }  // namespace ProcessLib
+
+#include "ThermoMechanicsFEM-impl.h"
```
Comment (endJunction) [inline ProcessLib/ThermoMechanics/ThermoMechanicsFEM.h]: Either all implementations go into the `impl.h` file, or no. I prefer the latter.

File: ProcessLib/ThermoMechanics/ThermoMechanicsProcess.cpp (hunk 4 above)
Comment (wenqing) [reply ProcessLib/ThermoMechanics/ThermoMechanicsProcess.cpp]: Corrected.

File: ProcessLib/ThermoMechanics/ThermoMechanicsFEM-impl.h (hunk 5 above)
Comment (wenqing) [reply ProcessLib/ThermoMechanics/ThermoMechanicsFEM-impl.h]: Changed.

File: ProcessLib/ThermoMechanics/ThermoMechanicsFEM.h (hunk 6 above)
Comment (wenqing) [reply ProcessLib/ThermoMechanics/ThermoMechanicsFEM.h]: For an easy reviewing, I did not move the existing function definition  to impl.h. I planned to do it once this PR is merged.

**Comment by wenqing** [comment]
@endJunction As discussed, I am going to merge this PR.

**Pull request merged**